### PR TITLE
[DO NOT MERGE] Attempt at keeping the DiveImportedModel consistent (contains #2297)

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [x] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

There were numerous ways in which the DiveImportedModel could go out of sync. For example by sharing a single dive table with the download thread(!). Also the import-functions consume the tables without the model being adapted. I see no obvious way how this can lead to crashes. But let's clean this up as a first step.

This also removes moving pointers via QML, which has us caused terrible problems before.

This is FYI. It is only compile tested. If some brave person compiles and tests download that would be very helpful (desktop and mobile). Currently, I can't test this as I haven't found time to install Bluetooth.

### Changes made:

Too many to list in detail, sorry.
1) Make DiveImportedModel the authoritative source of the imported dives.
2) Implement move semantics to move downloaded dives from DownloadThread to DiveImportedModel.
3) Keep model in sync.

### Related issues:

Related to #2292 and #2221, but probably doesn't fix them. They should give different backtraces, though.

### Mentions:

@dirkhh: FYI - can't work on this anymore today.

### CODE REVIEW

**Comment by bstoeger** [comment]
PS: In case you wonder - compiling the tests fails because core/downloadfromdcthread.cpp now directly accesses the DivesImportedModel.

The former is linked with the tests, the latter isn't. Either remove the thread from the core lib or link the tests with the models...

File: qt-models/diveimportedmodel.h [hunk 1]
```diff
@@ -4,8 +4,9 @@
 #include <QAbstractTableModel>
```
Comment (dirkhh) [inline qt-models/diveimportedmodel.h:4]: The subject line of the commit has a typo: `DiveImporedModel` should be `DiveImportedModel`
And the description in the commit message has two verbs "fill reload"

File: core/downloadfromdcthread.cpp [hunk 2]
```diff
@@ -63,6 +64,10 @@ DownloadThread::DownloadThread() : downloadTable({ 0 }),
 	diveSiteTable({ 0 }),
 	m_data(DCDeviceData::instance())
 {
+	// By connecting the finished signal in the constructor, we make sure that we are called first.
+	// Thus, we can update the model and all other recipients of the signal (notably QML) can
+	// check the model for new dives. Yes, this is somewhat subtle, but well defined.
+	connect(this, &QThread::finished, this, &DownloadThread::reloadModel);
```
Comment (dirkhh) [inline core/downloadfromdcthread.cpp:70]: I love this. Brilliant.

File: mobile-widgets/qml/DownloadFromDiveComputer.qml [hunk 3]
```diff
@@ -28,11 +28,6 @@ Kirigami.Page {
 		id: downloadThread
 
 		onFinished : {
-			if (!table || !sites) {
-				console.warn("DCDownloadThread::onFinished(): table or sites is null!")
-				return
-			}
-			importModel.repopulate(table, sites)
```
Comment (dirkhh) [inline mobile-widgets/qml/DownloadFromDiveComputer.qml:35]: not having to do this from QML is So. Much. Better.

**Review by dirkhh — APPROVED** [review]
Read through this twice and can see several spots where this prevents corruption that could happen with the current code. Thinking this through I'm now almost surprised that the current code works at all in many instances, as the correct tables were only available if the functions happened to be called in a certain order.
A couple of small comments. And I'll look at the build failure as well as I would love to see this merged so we can test this on Android (and iOS).

**Comment by dirkhh** [comment]
Hrmpf. The fix to compiling the tests seems straight forward. Just like for the regular desktop app, ad the `subsurface_models_desktop` library in `tests/CMakeLists.txt`
Except that fails with an undefined reference to `DiveImportedModel::repopulate()` - an issue we don't have when linking Subsurface - only with the tests.
Right now I'm feeling like I can't see the forest for the trees...

**Comment by bstoeger** [comment]
No worries - I will change it so that the model connects to the thread. Then things should compile again.

BTW: While I believe that the changes in this PR are in principle correct, I doubt that it will fix any crashes. The `data()` function was quite resilient against returning garbage. Let's see.

**Comment by dirkhh** [comment]
The latest Subsurface-mobile Android beta is built with this included - and one of the people who had crashes on import reports that he didn't see those crashes with this version

**Comment by bstoeger** [comment]
Interesting... I hope you realize that this was all done completely blind with the only testing being compile and run. 😮

**Comment by dirkhh** [comment]
> Interesting... I hope you realize that this was all done completely blind with the only testing being compile and run. 😮

I didn't. I reviewed the code and it seemed to make sense and I could identify two situations around download thread where this seems much more likely to do the right thing...

**Comment by bstoeger** [comment]
Ok - that's good news. At least we have a grasp of what we should never do:
1) Pass C-pointers through QML.
2) Modify models in QML context.

I have an idea how to solve this in a much nicer way, which will mean completely rewriting this PR. Please give me a few days - currently I have some mineralogical problems to tackle.

**Comment by bstoeger** [comment]
Yes, let's redo this in a more sensible way...

**Pull request closed without merging**